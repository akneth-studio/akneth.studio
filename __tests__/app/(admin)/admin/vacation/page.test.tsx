/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, fireEvent, waitFor, cleanup } from '@testing-library/react';
import VacationAdminPage from '@/app/(admin)/admin/vacation/page';

// Mock the Auth component
jest.mock('@/components/admin/auth', () => ({
    __esModule: true,
    default: ({ children }: { children: React.ReactNode }) => <>{children}</>,
}));

global.fetch = jest.fn();

const mockBanners = [
    {
        id: '1',
        mode: 'vacation',
        announce_from: '2025-08-01T10:00:00.000Z',
        date_start: '2025-08-10T00:00:00.000Z',
        date_end: '2025-08-20T00:00:00.000Z',
        visible: true,
    },
    {
        id: '2',
        mode: 'maintenance',
        announce_from: '2025-09-01T12:00:00.000Z',
        date_start: '2025-09-05T00:00:00.000Z',
        date_end: '2025-09-06T00:00:00.000Z',
        visible: false,
    },
];

describe('VacationAdminPage', () => {
    beforeEach(() => {
        (global.fetch as jest.Mock).mockClear();
    });

    afterEach(() => {
        cleanup();
    });

    it('matches snapshot', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({
            ok: true,
            json: async () => mockBanners,
        });
        const { asFragment } = render(<VacationAdminPage />);
        await waitFor(() => {
            expect(screen.getByText('Urlop')).toBeInTheDocument();
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders loading state initially', () => {
        (global.fetch as jest.Mock).mockImplementation(() => new Promise(() => {})); // Never resolves
        render(<VacationAdminPage />);
        expect(screen.getByText('Wczytywanie…')).toBeInTheDocument();
    });

    it('renders banners after successful fetch', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({
            ok: true,
            json: async () => mockBanners,
        });
        render(<VacationAdminPage />);
        await waitFor(() => {
            expect(screen.getByText('Urlop')).toBeInTheDocument();
            expect(screen.getByText('Prace serwisowe')).toBeInTheDocument();
        });
    });

    it('renders no banners message when fetch returns empty array', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({
            ok: true,
            json: async () => [],
        });
        render(<VacationAdminPage />);
        await waitFor(() => {
            expect(screen.getByText('Brak komunikatów w bazie.')).toBeInTheDocument();
        });
    });

    it('displays an error message if fetching banners fails', async () => {
        (global.fetch as jest.Mock).mockRejectedValueOnce(new Error('Network Error'));
        render(<VacationAdminPage />);
        await waitFor(() => {
            expect(screen.getByText(/Błąd pobierania komunikatów/i)).toBeInTheDocument();
        });
    });

    it('opens and closes the add modal', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({
            ok: true,
            json: async () => [],
        });
        render(<VacationAdminPage />);
        await waitFor(() => {
            fireEvent.click(screen.getByTestId('add-banner-button'));
        });
        expect(screen.getByTestId('modal-title')).toHaveTextContent('Dodaj komunikat');
        fireEvent.click(screen.getByText('Anuluj'));
        await waitFor(() => {
            expect(screen.queryByTestId('modal-title')).not.toBeInTheDocument();
        });
    });

    it('adds a new banner and shows success message', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => [] });
        render(<VacationAdminPage />);
        await waitFor(() => fireEvent.click(screen.getByTestId('add-banner-button')));

        fireEvent.change(screen.getByLabelText(/Początek/i), { target: { value: '2025-08-10T00:00' } });
        fireEvent.change(screen.getByLabelText(/Koniec/i), { target: { value: '2025-08-20T00:00' } });
        fireEvent.change(screen.getByLabelText(/Ogłoszenie od/i), { target: { value: '2025-08-01T10:00' } });

        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => mockBanners });

        fireEvent.click(screen.getByTestId('modal-submit-button'));

        await waitFor(() => {
            expect(global.fetch).toHaveBeenCalledWith('/api/banners', expect.objectContaining({ method: 'POST' }));
            expect(screen.getByText('Dodano komunikat.')).toBeInTheDocument();
        });
    });

    it('shows an error message when adding a banner fails', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => [] });
        render(<VacationAdminPage />);
        await waitFor(() => fireEvent.click(screen.getByTestId('add-banner-button')));

        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: false, json: async () => ({ error: { message: 'Failed to add' } }) });

        fireEvent.click(screen.getByTestId('modal-submit-button'));

        await waitFor(() => {
            expect(screen.getByText('Failed to add')).toBeInTheDocument();
        });
    });

    it('edits a banner and shows success message', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => mockBanners });
        render(<VacationAdminPage />);
        await waitFor(() => fireEvent.click(screen.getByTestId('edit-banner-1')));

        fireEvent.change(screen.getByLabelText(/Początek/i), { target: { value: '2025-08-11T00:00' } });

        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => ({}) });
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => mockBanners });

        fireEvent.click(screen.getByTestId('modal-submit-button'));

        await waitFor(() => {
            expect(global.fetch).toHaveBeenCalledWith('/api/banners', expect.objectContaining({ method: 'PATCH' }));
            expect(screen.getByText('Zaktualizowano komunikat.')).toBeInTheDocument();
        });
    });

    it('shows an error message when editing a banner fails', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => mockBanners });
        render(<VacationAdminPage />);
        await waitFor(() => fireEvent.click(screen.getByTestId('edit-banner-1')));

        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: false, json: async () => ({ error: { message: 'Failed to update' } }) });

        fireEvent.click(screen.getByTestId('modal-submit-button'));

        await waitFor(() => {
            expect(screen.getByText('Failed to update')).toBeInTheDocument();
        });
    });

    it('deletes a banner and shows success message', async () => {
        window.confirm = jest.fn(() => true);
        (global.fetch as jest.Mock)
            .mockResolvedValueOnce({ ok: true, json: async () => mockBanners }) // Initial fetch for rendering
            .mockResolvedValueOnce({ ok: true, json: async () => ({}) }) // Mock for the DELETE request
            .mockResolvedValueOnce({ ok: true, json: async () => [] }); // Mock for the re-fetch after delete
        render(<VacationAdminPage />);
        await screen.findByText('Urlop'); // Wait for banners to render
        const deleteButton = await screen.findByTestId('delete-banner-1');
        fireEvent.click(deleteButton);

        await waitFor(() => {
            expect(global.fetch).toHaveBeenCalledWith('/api/banners', expect.objectContaining({ method: 'DELETE' }));
        });
        expect(await screen.findByText(/Usunięto komunikat/i)).toBeInTheDocument();
    });

    it('shows an error message when deleting a banner fails', async () => {
        window.confirm = jest.fn(() => true);
        (global.fetch as jest.Mock)
            .mockResolvedValueOnce({ ok: true, json: async () => mockBanners }) // Initial fetch for rendering
            .mockResolvedValueOnce({ ok: false, json: async () => ({ error: { message: 'Failed to delete' } }) }); // Mock for the DELETE request
        render(<VacationAdminPage />);
        await screen.findByText('Urlop'); // Wait for banners to render
        const deleteButton = await screen.findByTestId('delete-banner-1');
        fireEvent.click(deleteButton);

        expect(await screen.findByText('Failed to delete')).toBeInTheDocument();
    });

    it('does not delete if user cancels confirmation', async () => {
        window.confirm = jest.fn(() => false);
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => mockBanners }); // Initial fetch for rendering
        render(<VacationAdminPage />);
        await screen.findByText('Urlop'); // Wait for banners to render
        const deleteButton = await screen.findByTestId('delete-banner-1');
        fireEvent.click(deleteButton);

        expect(global.fetch).toHaveBeenCalledTimes(1);
    });

    it('shows validation errors for empty fields', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => [] });
        render(<VacationAdminPage />);
        await waitFor(() => fireEvent.click(screen.getByTestId('add-banner-button')));

        fireEvent.change(screen.getByLabelText(/Początek/i), { target: { value: '' } });
        fireEvent.change(screen.getByLabelText(/Koniec/i), { target: { value: '' } });
        fireEvent.change(screen.getByLabelText(/Ogłoszenie od/i), { target: { value: '' } });

        const formElement = screen.getByTestId('modal-submit-button').closest('form');
        if (formElement) {
            fireEvent.submit(formElement);
        }

        await waitFor(async () => {
            expect(await screen.findAllByText(/Wymagane\./i)).toHaveLength(3);
        });
    });

    it('shows validation error for end date before start date', async () => {
        (global.fetch as jest.Mock).mockResolvedValueOnce({ ok: true, json: async () => [] });
        render(<VacationAdminPage />);
        await waitFor(() => fireEvent.click(screen.getByTestId('add-banner-button')));

        fireEvent.change(screen.getByLabelText(/Początek/i), { target: { value: '2025-08-20T00:00' } });
        fireEvent.change(screen.getByLabelText(/Koniec/i), { target: { value: '2025-08-10T00:00' } });

        const formElement = screen.getByTestId('modal-submit-button').closest('form');
        if (formElement) {
            fireEvent.submit(formElement);
        }

        await waitFor(async () => {
            expect(await screen.findByText('Data zakończenia nie może być wcześniejsza niż początek.')).toBeInTheDocument();
        });
    });
});