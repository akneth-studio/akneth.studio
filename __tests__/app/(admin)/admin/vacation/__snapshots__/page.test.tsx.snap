// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`VacationAdminPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="py-4"
  >
    <div
      class="d-flex justify-content-between align-items-center mb-4"
    >
      <h2
        class="fw-bold mb-0"
      >
        Zarządzanie komunikatami (urlop, serwis, historia)
      </h2>
      <button
        aria-label="Dodaj komunikat"
        class="btn btn-primary d-flex align-items-center btn btn-primary"
        data-testid="add-banner-button"
        role="button"
        tabindex="0"
        type="button"
      >
        <svg
          class="me-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 11h-6V5h-2v6H5v2h6v6h2v-6h6z"
          />
        </svg>
        Dodaj komunikat
      </button>
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-bordered table-hover"
      >
        <thead
          class="table-light"
        >
          <tr>
            <th>
              ID
            </th>
            <th>
              Typ
            </th>
            <th>
              Zakres
            </th>
            <th>
              Ogłoszenie
            </th>
            <th>
              Widoczny
            </th>
            <th>
              Akcje
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <span>
                2…
              </span>
            </td>
            <td>
              Prace serwisowe
            </td>
            <td>
              5.09.2025 —
              <br />
              6.09.2025
            </td>
            <td>
              1.09.2025, 14:00:00
            </td>
            <td>
              NIE
            </td>
            <td
              class="d-flex gap-2"
            >
              <button
                aria-label="Edytuj"
                class="btn btn-outline-primary d-flex align-items-center btn btn-primary"
                data-testid="edit-banner-2"
                role="button"
                tabindex="0"
                type="button"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 21a1 1 0 0 0 .24 0l4-1a1 1 0 0 0 .47-.26L21 7.41a2 2 0 0 0 0-2.82L19.42 3a2 2 0 0 0-2.83 0L4.3 15.29a1.06 1.06 0 0 0-.27.47l-1 4A1 1 0 0 0 3.76 21 1 1 0 0 0 4 21zM18 4.41 19.59 6 18 7.59 16.42 6zM5.91 16.51 15 7.41 16.59 9l-9.1 9.1-2.11.52z"
                  />
                </svg>
                Edytuj
              </button>
              <button
                aria-label="Usuń"
                class="btn btn-outline-danger d-flex align-items-center btn btn-primary"
                data-testid="delete-banner-2"
                role="button"
                tabindex="0"
                type="button"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5 20a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V8h2V6h-4V4a2 2 0 0 0-2-2H9a2 2 0 0 0-2 2v2H3v2h2zM9 4h6v2H9zM8 8h9v12H7V8z"
                  />
                  <path
                    d="M9 10h2v8H9zm4 0h2v8h-2z"
                  />
                </svg>
                Usuń
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <span>
                1…
              </span>
            </td>
            <td>
              Urlop
            </td>
            <td>
              10.08.2025 —
              <br />
              20.08.2025
            </td>
            <td>
              1.08.2025, 12:00:00
            </td>
            <td>
              TAK
            </td>
            <td
              class="d-flex gap-2"
            >
              <button
                aria-label="Edytuj"
                class="btn btn-outline-primary d-flex align-items-center btn btn-primary"
                data-testid="edit-banner-1"
                role="button"
                tabindex="0"
                type="button"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 21a1 1 0 0 0 .24 0l4-1a1 1 0 0 0 .47-.26L21 7.41a2 2 0 0 0 0-2.82L19.42 3a2 2 0 0 0-2.83 0L4.3 15.29a1.06 1.06 0 0 0-.27.47l-1 4A1 1 0 0 0 3.76 21 1 1 0 0 0 4 21zM18 4.41 19.59 6 18 7.59 16.42 6zM5.91 16.51 15 7.41 16.59 9l-9.1 9.1-2.11.52z"
                  />
                </svg>
                Edytuj
              </button>
              <button
                aria-label="Usuń"
                class="btn btn-outline-danger d-flex align-items-center btn btn-primary"
                data-testid="delete-banner-1"
                role="button"
                tabindex="0"
                type="button"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5 20a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V8h2V6h-4V4a2 2 0 0 0-2-2H9a2 2 0 0 0-2 2v2H3v2h2zM9 4h6v2H9zM8 8h9v12H7V8z"
                  />
                  <path
                    d="M9 10h2v8H9zm4 0h2v8h-2z"
                  />
                </svg>
                Usuń
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
